mpy-cross/mpy_cross/compiler.py

```diff
@@ -0,0 +1,204 @@
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import os
+
+
+class CrossCompileError(Exception):
+    pass
+
+
+class CompilerNotFoundError(Exception):
+    pass
+
+
+EMIT_BYTECODE = "bytecode"
+EMIT_NATIVE = "native"
+EMIT_VIPER = "viper"
+
+
+def emit_id(name):
+    return {
+        EMIT_BYTECODE: 1,
+        EMIT_NATIVE: 2,
+        EMIT_VIPER: 3,
+    }[name]
+
+
+NATIVE_ARCHS = {
+    "NATIVE_ARCH_NONE": "",
+    "NATIVE_ARCH_X86": "x86",
+    "NATIVE_ARCH_X64": "x64",
+    "NATIVE_ARCH_ARMV6": "armv6",
+    "NATIVE_ARCH_ARMV6M": "armv6m",
+    "NATIVE_ARCH_ARMV7M": "armv7m",
+    "NATIVE_ARCH_ARMV7EM": "armv7em",
+    "NATIVE_ARCH_ARMV7EMSP": "armv7emsp",
+    "NATIVE_ARCH_ARMV7EMDP": "armv7emdp",
+    "NATIVE_ARCH_XTENSA": "xtensa",
+    "NATIVE_ARCH_XTENSAWIN": "xtensawin",
+}
+
+globals().update(NATIVE_ARCHS)
+
+
+def native_arch_id(name):
+    if name is None:
+        return 0
+
+    return {
+        "": 0,
+        "x86": 1,
+        "x64": 2,
+        "armv6": 3,
+        "armv6m": 4,
+        "armv7m": 5,
+        "armv7em": 6,
+        "armv7emsp": 7,
+        "armv7emdp": 8,
+        "xtensa": 9,
+        "xtensawin": 10,
+    }[name]
+
+
+class Compiler:
+    def mpy_version(self):
+        """
+        Get the version and sub-version of the .mpy file format generated by this compiler.
+
+        Returns: A tuple of `(mpy_version, mpy_sub_version)`
+        """
+        raise NotImplementedError()
+
+    def version(self):
+        """
+        Get the version string from this compiler.
+
+        Returns: A string, typically "MicroPython <git tag> on <build date>".
+        """
+        raise NotImplementedError()
+
+    def compile(self, src, dest=None, src_path=None, opt=0, march=None, emit=EMIT_BYTECODE):
+        """
+        Compile the specified .py file with mpy-cross.
+
+        Returns: Standard output from mpy-cross as a string.
+
+        Required arguments:
+         - src:        The path to the .py file
+
+        Optional keyword arguments:
+         - dest:       The output .mpy file. Defaults to `src` (with .mpy extension)
+         - src_path:   The path to embed in the .mpy file (defaults to `src`)
+         - opt:        Optimisation level (0-3, default 0)
+         - march:      One of the `NATIVE_ARCH_*` constants (defaults to
+                       None, which is treated as NATIVE_ARCH_NONE). Architecture to
```

Comment:
Because of the silly tricks I was doing to declare the NATIVE_ARCH_* constants above, it was confusing the ruff linter. I've just made the constants explicitly, and updated it like you suggest.